---
title: "Single-cell Bioinformatics Analysis"
author: | 
  | Mengrui Zhang, Magdy Alabady
  | The Georgia Genomics and Bioinformatics Core (GGBC)
date: "8/23/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Introduction

In this analysis, we will perform the analysis of single-cell RNAseq data on Cat. The data were from 10x genomics analysis pipeline by using "Cellranger counts" with the cat reference genome. The R packages we use in this analysis will be "Monocle3" and "Seurat". 

We plan to perform the following:

1. Preprocessing the data using Principal components analysis (PCA). 
2. Cluster cells using UMAP and Louvain methods. 
3. Construct cell trajectories. 
4. Oder cells in chronological order. 
5. Pseudo time analysis. 
6. Significant gene analysis. 

# 2. Data set analysis

## 2.1 Raw data analysi using "Cellranger" and "10x Loupe Cell Browser"

In this section, for the first step, we run an analysis pipeline using "Cellranger count" by 10x genomics. The reference genome was created using "Felis_catus_9.0" database. The result can be shown by using the "10x Loupe Cell Browser". The following 2 figure is an example of "10x Loupe Cell Browser" interface. 


```{r echo=FALSE, out.width='70%'}
knitr::include_graphics('C:/Users/mengr/Dropbox/GGBC/cat/data_analysis/graph/loupe.png')
```

This figure shows the clustering result (K-mean) using "10x Loupe Cell Browser". 

```{r echo=FALSE, out.width='70%'}
knitr::include_graphics('C:/Users/mengr/Dropbox/GGBC/cat/data_analysis/graph/loupe2.png')
```

The above figure shows the clustering result from "Loupe" using the t-SNE method. There are more detailed results inside of the "10x Loupe Cell Browser", the software has a great interactive interference for users. 

## 2.2 Load dataset into R package "Monocle3" and preprocessing with PCA

The output of Cellranger analysis pipeline has three files: "barcodes", "features" and "matrix". The "barcodes" file contains all the cell names and the "feature" file contain the genes and gene short names. The "matrix" file includes the counts with the location of gene and cell. We load this data into Rstudio. 

```{r include=FALSE}
library(monocle3)
library(reticulate)
library(stringr)
library(Seurat)
library(ggplot2)
library(knitr)
library(kableExtra)
library(pander)
library(dplyr)
##### Read 10x Data ##############
# Provide the path to the Cell Ranger output.
cat <- load_cellranger_data("C:/Users/mengr/Dropbox/GGBC/cat/data_analysis", barcode_filtered = TRUE)
```

```{r echo=FALSE}
cat
```

The above is a summary of the R object for this dataset. There is a total of 10000 cells and 26373 genes. We preprocessing the data into 100 dimensions for our later analysis. This preprocessing step is to normalize the data, to use Principal Components Analysis (the standard for RNA-seq) and to remove any batch effects.

```{r echo=FALSE}
cat = preprocess_cds(cat, method = "PCA", num_dim = 100)
plot_pc_variance_explained(cat)
```
Here we plot the PCA result with 100 dimensions. We can see the importance of component decrease as the dimension increase. We choose 100 dimensions for our data should be enough. 

## 2.3 Cluster cells using UMAP and Louvain methods.

We cluster the cells using Uniform Manifold Approximation and Projection (UMAP). There is a total of 10 clusters. 

```{r include=FALSE}
#Umap-PCA
cat = reduce_dimension(cat, reduction_method="UMAP", preprocess_method = "PCA")

cat = cluster_cells(cat, resolution = 1e-4, verbose = T,
                    python_home = "C:/Users/mengr/location/.conda/envs/r-reticulate/python")
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
plot_cells(cat, reduction_method="UMAP")
```

## 2.4 cluster the cells using tSNE

We cluster the cells using T-distributed Stochastic Neighbor Embedding(tSNE).

```{r include=FALSE}
cat = reduce_dimension(cat, reduction_method="tSNE", preprocess_method = "PCA")
cat = cluster_cells(cat, reduction_method="tSNE", resolution = 1e-4, verbose = F,
                    python_home = "C:/Users/theuser/location/.conda/envs/r-reticulate/python")
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
plot_cells(cat, reduction_method="tSNE")
```

This figure shows the clustering result using "t-SNE" method. 

## 2.5. Find Marker gene for each cluster. 

```{r include=FALSE}
marker_test_res = top_markers(cat, group_cells_by="cluster", reference_cells=1000, cores=8)
```
```{r echo=FALSE, warning=FALSE}
top_specific_markers = marker_test_res %>%
    filter(fraction_expressing >= 0.10) %>%
    group_by(cell_group) %>%
    top_n(3, pseudo_R2)
top_specific_marker_ids = unique(top_specific_markers %>% pull(gene_id))
```

Now we run a test to find out what genes makes them different from one another. Then we can rank markers genes according to "pseudo R2" value. For each cluster, we choose top 3 genes according to the "pseudo R2" value. The following table shows a total of 30 genes for all 10 clusters by UMAP.  

```{r echo=FALSE, warning=FALSE}
pander::pander(top_specific_markers[order(top_specific_markers$cell_group, decreasing = TRUE),])
```

```{r echo=FALSE, warning=FALSE}
plot_genes_by_group(cat,
                    top_specific_marker_ids,
                    group_cells_by="cluster",
                    ordering_type="maximal_on_diag",
                    max.size=3)
```

The above figure is a heatmap shows the significant genes and their corresponding clusters. 


## 2.6 Construct trajectories

The trajectories of cells can be learned by using "monocle3" package. The trajectories were calculated based on the UMAP cluster method.

```{r include=FALSE}
cat <- learn_graph(cat)
```

```{r echo=FALSE}
plot_cells(cat, reduction_method="UMAP")
```

From the above figure, we can see that there are about 2 major trajectories on the graph, one trajectory on the left goes from top to bottom and one trajectory on the center which has around 5 branches. There are also a few small trajectories which a little far from the major clusters. 

## 2.7 Oder cells in chronological order

```{r include=FALSE}
# Load the dataset
cat.data <- Read10X(data.dir = "C:/Users/mengr/Dropbox/GGBC/cat/data_analysis/outs/filtered_feature_bc_matrix")

#seurat normalization
cat_sc <- NormalizeData(cat.data, normalization.method = "LogNormalize", scale.factor = 10000)
cat_data <- as.data.frame(cat_sc)
cat_data$gene <- rownames(cat_data); rownames(cat_data) <- 1:nrow(cat_data)
```


In this section, We first use the R package "Seurat" basically to create a cell-gene data matrix and perform normalization. The following table is the significant genes in each of the developmental stages. 

```{r include=FALSE}
# Read marker genes
marker <- read.csv("C:/Users/mengr/Dropbox/GGBC/cat/data_analysis/marker.csv", stringsAsFactors = F)
marker <- marker[,1:5]
```

```{r echo=FALSE, warning=FALSE}
pander::pander(marker[,3:5])
```

In this analysis, we select cells that are in the first stage (Spermatogonial Stem Cells (SSC's)-Stem Progenitor) and then construct the pseudo time analysis. By using the significant genes, we can find significant cells that contain significant genes. We will make a cutoff point here to subset the cells. The next step is to find the nearest nodes(points) on the trajectories for the significant cells. We order those nodes and construct the pseudo time analysis. 

```{r include=FALSE}
cmlu <- str_split_fixed(marker$Genes, "- ", n = Inf)
cmlu[cmlu==""] <- NA
#cmlu <- na.omit(cmlu)

gene_names = cmlu[1,]
gene_names <- na.omit(gene_names)
```

\begin{table}[ht]
\centering
\begin{tabular}{rl}
  \hline
 & gene\_names \\ 
  \hline
1 & ID4 \\ 
  2 & HSPA8 \\ 
  3 & TAF4B \\ 
  4 & PEG10 \\ 
  5 & RBM5 \\ 
  6 & TCN2 \\ 
  7 & ISOC1 \\ 
  8 & HSD17B14 \\ 
  9 & PHOSPHO2 \\ 
  10 & ENPP2 \\ 
  11 & PRDX2 \\ 
  12 & ANXA1 \\ 
  13 & SERPINA5 \\ 
   \hline
\end{tabular}
\end{table}

The above figure includes all the gene short names at the stage: Spermatogonial Stem Cells (SSC's)-Stem Progenitor. We choose the starting points based on these genes. 

```{r include=FALSE}
## Subset matrix
sub_gene <- data.frame()
for (jj in 1:length(gene_names)){
  subset_data <- cat_data[cat_data$gene == gene_names[jj],]
#  subset_data <- subset_data[ , -which(names(subset_data) %in% c("gene"))]
  sub_gene <- rbind(sub_gene, subset_data)
}
gene = sub_gene$gene
sub_gene <- sub_gene[ , -which(names(sub_gene) %in% c("gene"))]
t1 = as.matrix(t(sub_gene))
t1 = as.vector(t1)
tt1 = data.frame(counts = t1, gene = rep(gene, each = 10000))
t2 = tt1[tt1$counts!=0,]
#x = seq(1,length(t2))
```

```{r echo=FALSE}
#boxplot(t2)
#tt = as.data.frame(t2)
g <- ggplot(t2, aes(t2$gene,t2$counts))
g + geom_boxplot(varwidth=T, fill="plum") + 
    labs(title="Box plot", 
         subtitle="Normalized counts distribution in Significant genes",
         caption="Source: Cat",
         x="Significant Genes",
         y="Log-normalized Counts")

```

The figure shows a box plot of all the significant genes in stage 1 and their counts after log-normalization. Now, we choose a cutoff point as 2.8 (can be discussed).

```{r include=FALSE}
#########
string_cell_name <- c()
cutoff = 2.8

for (ii in 1:length(gene_names)){
  subset_data <- cat_data[cat_data$gene == gene_names[ii],]
  subset_data <- subset_data[ , -which(names(subset_data) %in% c("gene"))]
  
  temp <- colnames(subset_data[,subset_data >= cutoff])
  temp2 <- paste(temp, "-1",sep = "")
  string_cell_name <- c(string_cell_name,temp2)
  print(ii)
}
```

```{r include=FALSE}
#########
string_cell_name <- c()
cutoff = 2.8

for (ii in 1:length(gene_names)){
  subset_data <- cat_data[cat_data$gene == gene_names[ii],]
  subset_data <- subset_data[ , -which(names(subset_data) %in% c("gene"))]
  
  temp <- colnames(subset_data[,subset_data >= cutoff])
  temp2 <- paste(temp, "-1",sep = "")
  string_cell_name <- c(string_cell_name,temp2)
  print(ii)
}

string = unique(string_cell_name)

get_earliest_principal_node <- function(cds,cell_ID){
  node_name = as.data.frame(igraph::V(principal_graph(cds)[["UMAP"]])$name)
  closest_vertex <- cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.data.frame(closest_vertex)
  closest_vertex$cell <- rownames(closest_vertex); rownames(closest_vertex) <- 1:nrow(closest_vertex)
  names(closest_vertex) <- c("node","cell")
  index <- closest_vertex[closest_vertex$cell %in% cell_ID,]$node
  root_nodes <- as.character(node_name[index,])
  root_nodes
}

```

```{r include=FALSE}
start <- get_earliest_principal_node(cat,string)
start <- unique(start)
cat <- order_cells(cat, root_pr_nodes=start)
```


## 2.8 Pseudo time analysis
```{r echo=FALSE}

plot_cells(cat,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5,
           cell_size = 0.8)
```

Here is the pseudo time analysis figure with the starting points as the significant genes from stage 1 (Spermatogonial Stem Cells (SSC's)-Stem Progenitor). 



